src/nomad_measurements/utils.py

```diff
@@ -153,3 +160,304 @@ def get_bounding_range_2d(ax1, ax2):
         ]
 
     return ax1_range, ax2_range
+
+
+class HDF5Handler:
+    """
+    Class for handling the creation of auxiliary files to store big data arrays outside
+    the main archive file (e.g. HDF5, NeXus).
+    """
+
+    def __init__(
+        self,
+        filename: str,
+        archive: 'EntryArchive',
+        logger: 'BoundLogger',
+        valid_dataset_paths: list = None,
+        nexus: bool = False,
+    ):
+        """
+        Initialize the handler.
+
+        Args:
+            filename (str): The name of the auxiliary file.
+            archive (EntryArchive): The NOMAD archive.
+            logger (BoundLogger): A structlog logger.
+            valid_dataset_paths (list): The list of valid dataset paths.
+            nexus (bool): If True, the file is created as a NeXus file.
+        """
+        if not filename.endswith(('.nxs', '.h5')):
+            raise ValueError('Only .h5 or .nxs files are supported.')
+
+        self.data_file = filename
+        self.archive = archive
+        self.logger = logger
+        self.valid_dataset_paths = []
+        if valid_dataset_paths:
+            self.valid_dataset_paths = valid_dataset_paths
+        self.nexus = nexus
+
+        self.hdf5_datasets = collections.OrderedDict()
+        self.hdf5_attributes = collections.OrderedDict()
+
+    def add_dataset(  # noqa: PLR0913
+        self,
+        path: str,
+        data: Any,
+        archive_path: str = None,
+        internal_reference: bool = False,
+        validate_path: bool = True,
+        lazy: bool = True,
```

Comment:
is there any case where this option is helpful? you can simply create an instance of the handler with one dataset and write it.

Reply:
I planned it such that only one instance of handler exists per normalization. In that case, if a dataset needs to be added right away because it needs to be read immediately, this could be helpful. 
On the other hand, the `read_dataset` method triggers the write functions anyways, so we don't need this flag here.

Reply:
I think you have one too many args, because if i understand it right when lazy you only add, so essentially you can get away with lazy by just using only add then if you like to write them use write. alternatively you can rid of add then have write take lazy arg so it skips actual writing when lazy.

Reply:
I agree, we can get rid of the lazy args. One can always trigger `write_file` once many datasets have been added. Or use the `write_file` right after adding one dataset